py/dml/serialize.py

```diff
@@ -372,80 +393,158 @@ def type_signature(dmltype, is_for_serialization):
                       if not is_for_serialization else '')
     assert False
 
+def serialize_sources_to_list(site, sources, out_attr):
+    sources = tuple(sources)
+    size = len(sources)
+    attr_alloc_expr = apply_c_fun(
+        site, "SIM_alloc_attr_list",
+        [ctree.mkIntegerConstant(site, size, False)],
+        attr_value_t)
+    attr_assign_statement = ctree.mkAssignStatement(
+        site, out_attr, ctree.ExpressionInitializer(attr_alloc_expr))
+    imm_attr_decl, imm_attr_ref = declare_variable(
+        site, "_imm_attr", attr_value_t)
+    statements = []
+    for (i, source_info) in enumerate(sources):
+        index = ctree.mkIntegerConstant(site, i, False)
+        if source_info is None:
+            sub_serialize = ctree.mkInline(site,
+                                           '_imm_attr = SIM_make_attr_nil();')
+        else:
+            (source, typ) = source_info
+            if typ is not None:
+                sub_serialize = serialize(typ, source, imm_attr_ref)
+            else:
+                sub_serialize = ctree.mkAssignStatement(
+                    site, imm_attr_ref, ctree.ExpressionInitializer(source))
+        sim_attr_list_set_statement = call_c_fun(
+            site, "SIM_attr_list_set_item", [ctree.mkAddressOf(site, out_attr),
+                                             index,
+                                             imm_attr_ref])
+        statements += [sub_serialize, sim_attr_list_set_statement]
+    ctree.mkCompound(
+        site, [attr_assign_statement, imm_attr_decl] + statements).toc()
```

Comment:
moving the `.toc()` to caller would make this function pure

Reply:
The fact `.toc()` of statements are used to implement this is an implementation detail, just like everything else in serialize.py.